File: Edge.m
%% Edge Class Definition
% Instances of the Edge class are used to define connections between nodes
% within a system.
%
% 1-May 2011
% Paul Grogan, ptgrogan@mit.edu
%%

classdef Edge < handle
    properties
        id;                 % unique identifier of edge
        origin;             % origin node handle
        destination;        % destination node handle
        type;               % edge type handle
        directed;           % flag for whether edge is directed or not
    end
    methods
        %% Edge Constructor
        % Instantiates a new Edge object with specified origin,
        % destination, type, and directed-ness.
        %
        % obj = Edge(id, origin, destination, type, directed)
        %   id:                 unique identifier of edge
        %   origin:             origin node handle
        %   destination:        destination node handle
        %   type:             	edge type handle
        %   directed:           flag (0 or 1) if edge is directed
        %
        % obj = Edge(origin, destination, edgeType, directed)
        %   origin:             origin node handle
        %   destination:        destination node handle
        %   type:             	edge type handle
        %   directed:           flag (0 or 1) if edge is directed
        %
        % obj = Edge()
        function obj=Edge(varargin)
            if nargin==5
                obj.id = varargin{1};
                obj.origin = varargin{2};
                obj.destination = varargin{3};
                obj.type = varargin{4};
                obj.directed = varargin{5};
            elseif nargin==4
                obj.id = CityNet.instance().GetNextEdgeId();
                obj.origin = varargin{1};
                obj.destination = varargin{2};
                obj.type = varargin{3};
                obj.directed = varargin{4};
            else
                obj.id = CityNet.instance().GetNextEdgeId();
                obj.origin = [];
                obj.destination = [];
                obj.type = [];
                obj.directed = 0;
            end
        end
        
        %% GetEuclideanLength Function
        % Gets the length of an edge using Euclidean distance metric.
        %
        % out = obj.GetEuclideanLength()
        %   out:    the length
        %   obj:    the edge object handle
        function out = GetEuclideanLength(obj)
            out = sqrt(sum((obj.origin.cell.location-obj.destination.cell.location).^2));
        end
        
        %% GetEdgeTypeAttributeValue Function
        % Accesses the edte type attribute value of a specific name.
        %
        % val = obj.GetEdgeTypeAttributeValue(attributeName)
        %   val:            the edge type attribute value
        %   obj:            the edge object handle
        %   attributeName:  the name of the attribute, string
        function val = GetEdgeTypeAttributeValue(obj,attributeName)
            if ~isempty(obj.type.attributes) ...
                && sum(strcmpi({obj.type.attributes.name},attributeName))==1
                val = obj.type.attributes( ...
                    strcmpi({obj.type.attributes.name},attributeName)).value;
            else
                val = [];
            end
        end
    end
end

File: java/src/edu/mit/citynet/core/Edge.java
/*
 * CityNet: Integrated Urban Development Decision Support Tool
 * 
 * Copyright (c) 2011 MIT Strategic Engineering Research Group
 */
package edu.mit.citynet.core;

import com.vividsolutions.jts.geom.LineSegment;

/**
 * The Edge class is used to define connections between nodes within a system.
 * 
 * @author Paul Grogan, ptgrogan@mit.edu
 */
public class Edge {
	private int id;
	private Node origin, destination;
	private EdgeType edgeType;
	private boolean directed;
	
	/**
	 * Gets the id.
	 *
	 * @return the id
	 */
	public int getId() {
		return id;
	}
	
	/**
	 * Sets the id.
	 *
	 * @param id the new id
	 */
	public void setId(int id) {
		this.id = id;
	}
	
	/**
	 * Gets the origin.
	 *
	 * @return the origin
	 */
	public Node getOrigin() {
		return origin;
	}
	
	/**
	 * Sets the origin.
	 *
	 * @param origin the new origin
	 */
	public void setOrigin(Node origin) {
		this.origin = origin;
	}
	
	/**
	 * Gets the destination.
	 *
	 * @return the destination
	 */
	public Node getDestination() {
		return destination;
	}
	
	/**
	 * Sets the destination.
	 *
	 * @param destination the new destination
	 */
	public void setDestination(Node destination) {
		this.destination = destination;
	}
	
	/**
	 * Gets the edge type.
	 *
	 * @return the edge type
	 */
	public EdgeType getEdgeType() {
		return edgeType;
	}
	
	/**
	 * Sets the edge type.
	 *
	 * @param edgeType the new edge type
	 */
	public void setEdgeType(EdgeType edgeType) {
		this.edgeType = edgeType;
	}
	
	/**
	 * Checks if is directed.
	 *
	 * @return true, if is directed
	 */
	public boolean isDirected() {
		return directed;
	}
	
	/**
	 * Sets the directed.
	 *
	 * @param directed the new directed
	 */
	public void setDirected(boolean directed) {
		this.directed = directed;
	}
	
	/**
	 * Gets the Euclidean length.
	 *
	 * @return the Euclidean length
	 */
	public double getEuclideanLength() {
		if(origin != null && destination != null) {
			LineSegment ls = new LineSegment(
					origin.getCell().getPolygon().getCentroid().getCoordinate(),
					destination.getCell().getPolygon().getCentroid().getCoordinate());
			return ls.getLength();
		} else return 0;
	}
	
	/**
	 * Gets the edge type attribute value.
	 *
	 * @param attributeName the attribute name
	 * @return the edge type attribute value
	 */
	public Double getEdgeTypeAttributeValue(String attributeName) {
		for(EdgeTypeAttribute attribute : edgeType.getAttributes()) {
			if(attribute.getName().equals(attributeName))
				return attribute.getValue();
		}
		return null;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	public boolean equals(Object object) {
		if(object instanceof Edge) {
			return id==((Edge)object).getId();
		} else return false;
	}
	
	/* (non-Javadoc)
	 * @see java.lang.Object#toString()
	 */
	public String toString() {
		return "Edge" + id;
	}
}
